include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +37,89 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(std::uintptr_t));
+    static_assert(sizeof(sycl::queue) % sizeof(std::uintptr_t) == 0);
+    static_assert(alignof(sycl::queue) >= alignof(std::uintptr_t));
+
+    std::uintptr_t __buf[sizeof(sycl::queue)/sizeof(std::uintptr_t)];
+
+    bool __has_queue() const
+    {
+        bool res = (__buf[0] != 0); // If the first size-of-pointer bytes are zeros, we consider there is no valid queue
+        std::atomic_signal_fence(std::memory_order_acq_rel); // to prevent possible reordering due to type punning
```

Comment:
For my own understanding, can you explain why we need this protection against reordering here?

Could the compiler not realize that `__buf[0]` and the return of `__queue_ref()` are in fact the same and lose some inherent protections against reordering?

Reply:
> Could the compiler not realize that __buf[0] and the return of __queue_ref() are in fact the same and lose some inherent protections against reordering?

Exactly. The C++ standard claims program behavior undefined if an object is accessed through an lvalue of an unrelated type (see [basic.lval] for details). That essentially allows compilers to assume "such access does not happen" and, in particular, do not even attempt any dependency analysis.